### DISCUSSION TITLE
Input required from project developer (I am planning on cleaning and improving the code)

### DISCUSSION THREAD

**Discussion by mp035** [RaspAP/raspap-webgui#735] [General]
Hi @billz ,
I am using raspap-webgui in a project and I will be adding some custom pages (which are irrelevant to the raspap-webgui project).  As part of this work I am planning on making the following 2 changes to the code which may be useful to your project:

1. Converting to conventional routing syntax (ie. http://raspberrpi.local/dashboard instead of http://raspberrypi.local?page=dashboard).  I have already completed this change, it does not require any external libraries, it just adds a single file to the lighttpd configuration folder.  I have added functions to the installer and uninstaller to handle this extra file.
2. Converting the html templates to BladeOne views ([https://github.com/EFTEC/BladeOne](https://github.com/EFTEC/BladeOne)) for easier maintenance and extension.  BladeOne is an ultra-light version of Laravel's blade engine.  It only requires a single php file, and does not appear to have any external dependencies.  This seems to fit with the current nature of raspap-webgui which doesn't appear to use composer (I assume because that would require yet another dependency on the Pi).  This does however introduce a vendor library, which may warrant adding a vendor folder and keeping in mind that the project should probably transition to PSR-4 standard sometime in the (distant?) future.

I wanted to present these two changes to you to see if you want them submitted to your project?  I will be doing the work anyway, but you may have good reasons for not including them, so let me know your thoughts.

Kind Regards, 
Mark

**Comment by billz** [comment]
Sounds great, Mark.

Updating the current routing scheme would be a nice improvement. This hasn't changed since the project's inception, for no reason other than historical inertia. If using lighttpd's mod_rewrite (or similar) gives us cleaner URLs I'm all for it.

Re: templating, the `renderTemplate()` function separates most of the logic from presentation, but there's still a fair bit of PHP in the project templates. One consideration is the  gettext() alias `_()` that's currently used for labels in these files. Do you know how BladeOne handles this? I assume the labels could be set with gettext() and passed as vars.

Until now RaspAP hasn't relied on external PHP dependencies (apart from Parsedown, a markdown parsing class). Certainly not opposed to it, or adding a /vendor library if it's needed and indeed eventually moving to PSR-4.

Give the branch `feature/dhcp-eth0` a look. This refactors RaspAP's internal DHCP config handling, with the goal of merging manual user configurations in a predictable way. It mainly touches /includes, but some templates and the installer are updated as well. I plan to merge this and create a 2.6-beta pre-release after a final round of testing. You might do a rebase against your fork to get a scope of potential conflicts, and decide if it makes sense for your project.

Thanks,
BZ

**Reply by mp035** [reply]
Hi BZ,
All points noted.  Blade syntax handles gettext pretty easily using {{ _($varToTranslate) }}.  It also allows php code in the template where needed (unlike twig).  I'll take a look at the dhcp-eth0 branch and make a call.

**Reply by mp035** [reply]
Ok, so I have rebased my working branch off of dhcp-eth0.  My thinking is that the changes will be kind of major, so it should be easier to handle pull requests to a non-master branch.

**Reply by mp035** [reply]
Hi @billz ,
After rebasing off of dhcp-eth0 the application no longer works in AP-STA mode.  From a fresh raspian image I can get the webgui up and runnnig but as soon as I select Enable Wifi Client Ap Mode and hit save settings, the wifi interface dies, and can not be restarted.  I tried manually running the /etc/raspap/hostapd/servicestart.sh script to see what happens, but I'm a bit out of my comfort zone here.  The service fails to start, and attached is ac copy of journalctl.

[journalctl.txt](https://github.com/billz/raspap-webgui/files/5707661/journalctl.txt)

Any ideas where to start looking?

Cheers.

**Comment by billz** [comment]
> dhcpcd[2492]: wlan0: IAID conflicts with one assigned to uap0

Default dhcp + dnsmasq values that were previously hardcoded for these interfaces now live in `config/defaults.json`. This file is moved by the installer to `/etc/raspap/networking`. If the installer was not run with the `--branch` option these defaults are probably missing. There are some updates to `raspap.sudoers` that will need to be applied as well.

**Reply by mp035** [reply]
The installer was run with the -b option pointing to my rebased branch to make sure the changes to lighttpd config files were made correctly (which they appear to be).  So I double checked defaults.json was correctly installed, and it seems to be right,
Here's the console transcript with sdcard mounted on my manjaro laptop.

![image](https://user-images.githubusercontent.com/9688384/102481072-f2fab000-40ac-11eb-92ba-054c1f946034.png)

**Reply by mp035** [reply]
Found it.  /etc/hostapd/hostapd.conf had `interface=wlan0` and it should have been `interface=uap0`.  I changed it and it is now working.

I will re-run the install procedure, taking notice of the interface value as I set it up, because I wonder if it is another 'band-aid' lurking in the code that selects the wrong interface.

Thanks heaps for the assistance you have given me.

**Comment by billz** [comment]
No worries, glad to have this collaboration. 

Re: hostapd.conf, I've identified the bug and pushed a fix. Thanks for catching it.

**Reply by mp035** [reply]
BTW your fix has a typo in your master branch.  includes/hostapd.php line 359:
`function updateHostapdConfig($ignore_broadcast_ssid,$wifiAPEnble,$bridgedEnable)`
should be
`function updateHostapdConfig($ignore_broadcast_ssid, $wifiAPEnable, $bridgedEnable)`
(you forgot the **a** in $wifiApEn**a**ble)

I'll push a commit to my latest pull request to fix it in dhcp-etho branch.

**Reply by billz** [reply]
Thanks for spotting this. I've pushed a fix to the master branch.

**Reply by mp035** [reply]
I Just noticed the same problem on line 251 of the same file.
